### PR TITLE
Simplify add cluster view

### LINKED ISSUE #2568: Improve Add Cluster UI
**What would you like to be added**:
Switch the selection of context to a list of labelled toggles. In conjunction it should display contexts that are invalid with errors in line (and the toggle itself being disabled).

**Why is this needed**:

Currently, when selecting a kubeconfig the contexts are detected and put into a dropdown. However, this makes for less than ideal UX since it requires 2 addtional clicks (one to open the drop down and one to close it) whenever selecting context(s).

### PR DESCRIPTION
**Pull request by Nokel81**
- Only shows editor, and will save all text to file

- Reports some validation errors, but doesn't block the add cluster button

- Cleanup kube helpers to facilitate clearer FS interactions

Signed-off-by: Sebastian Malton <sebastian@malton.name>

### CODE REVIEW

File: src/common/kube-helpers.ts [hunk 1]
```diff
@@ -71,67 +131,44 @@ export function loadFromOptions(options: KubeConfigOptions): KubeConfig {
   return kc;
 }
 
-export function loadConfig(pathOrContent?: string): KubeConfig {
-  return loadConfigFromString(
-    fse.pathExistsSync(pathOrContent)
-      ? readResolvedPathSync(pathOrContent)
-      : pathOrContent
-  );
+export interface ConfigResult {
```
Comment (Nachasic) [inline src/common/kube-helpers.ts:155]: Sounds like a good precedent to adopt monadic return types (e.g. https://github.com/hqoss/monads) in the future? This particular case looks like it would be a good use of `Result<KubeConfig, Joi.ValidationError>`

File: src/common/kube-helpers.ts [hunk 2]
```diff
@@ -71,67 +131,44 @@ export function loadFromOptions(options: KubeConfigOptions): KubeConfig {
...
+  const { options, error } = loadToOptions(content);
 
-  if (!config.clusters || config.clusters.length == 0) {
-    throw new Error("No clusters provided in config");
-  }
-
-  if (!config.contexts || config.contexts.length == 0) {
-    throw new Error("No contexts provided in config");
-  }
+  return {
```
Comment (Nachasic) [inline src/common/kube-helpers.ts:163]: Monadic types (referring to my previous comment) could be mapped:
```ts
// assuming `loadToOptions()` returns `Result<Options, Error>`
return loadToOptions(content).map(options => loadFromOptions(options)) // Result<Options, Error>
```

File: src/main/catalog-sources/kubeconfig-sync.ts [hunk 3]
```diff
@@ -127,7 +125,13 @@ type RootSource = ObservableMap<string, RootSourceValue>;
 export function computeDiff(contents: string, source: RootSource, filePath: string): void {
   runInAction(() => {
     try {
-      const rawModels = configToModels(loadConfigFromString(contents), filePath);
+      const { config, error } = loadConfigFromString(contents);
```
Comment (Nachasic) [inline src/main/catalog-sources/kubeconfig-sync.ts:157]: Again, with monadic types this could be simplified to 
```ts
const config = loadConfigFromString(contents);

if (config.isErr()) {
    logger.warn();
}
const rawModels = configToModels(config.unwrapOr(fallbackConfig), filePath);
```

**Review by Nachasic — DISMISSED** [review]
Posted some comments for consideration, LGTM

File: src/common/kube-helpers.ts (hunk 1 above)
Comment (Nokel81) [reply src/common/kube-helpers.ts:155]: While I really like monadic return types (especially coming from my liking of rust). I don't think this place would be good. Because it really should be (if I was in rust) `(KubeConfig, Option<Joi.ValidationError>)`, which isn't that more ergonomic than what I have now.

Comment (Nachasic) [reply src/common/kube-helpers.ts:155]: Makes sense.

**Comment by aleksfront** [comment]
I have couple of question related comparing before/after views:

Before
current view

After
simplified view

1. Is it simplifies view or some interactions in it? Not much elements changed so it's still "simple" (maybe not simple for the user, however)
2. Should we remove description which is currently presented ("Paste kubeconfig as a text...")?
3. New message `Clusters added here are **not** merged...` doesn't seem right. *Not* word can confuse user. For example, first thoughts for me was "why new clusters should be merged?"

From the UI point of view I suggest whole look and UX should be refactored here. But we can do this in separate PR though.

**Comment by Nokel81** [comment]
@aleksfront 
1. It simplifies it because there was initially two PRs (one for a fix like this and one to add the ability to add AccessibleNamespaces). But now there is just this one. I don't just simplify the code and the top text, but it also removes the "Proxy Settings"
2. I have removed that description, because we don't support that anymore.
3. I have that message because when you use `kubectl config set-context` (and siblings) it does modify the default kubeconfig file (unless you specify the `--config` option).

**Review by Nachasic — DISMISSED** [review]

File: package.json [hunk 4]
```diff
@@ -181,6 +181,7 @@
     "@hapi/call": "^8.0.0",
     "@hapi/subtext": "^7.0.3",
     "@kubernetes/client-node": "^0.12.0",
+    "@material-ui/icons": "^4.11.2",
```
Comment (jakolehm) [inline package.json]: Is this actually used?

File: src/common/utils/index.ts [hunk 5]
```diff
@@ -21,7 +21,9 @@
 
 // Common utils (main OR renderer)
 
-export const noop: any = () => { /* empty */ };
+export function noop<T extends any[]>(...args: T): void {
```
Comment (jakolehm) [inline src/common/utils/index.ts:24]: Looks like unnecessary change? Is it?

File: src/renderer/bootstrap.tsx [hunk 6]
```diff
@@ -44,11 +44,11 @@ import { HelmRepoManager } from "../main/helm/helm-repo-manager";
 import { ExtensionInstallationStateStore } from "./components/+extensions/extension-install.store";
 import { DefaultProps } from "./mui-base-theme";
 
-/**
```
Comment (jakolehm) [inline src/renderer/bootstrap.tsx]: Any reason for removing this?

File: package.json (hunk 4 above)
Comment (Nokel81) [reply package.json]: Ah this is actually a duplicate of the item in `devDependencies`

File: src/common/utils/index.ts (hunk 5 above)
Comment (Nokel81) [reply src/common/utils/index.ts:24]: It is a minor improvement in the typing. That is all. I forget why I did this in the original PRs.

File: src/renderer/bootstrap.tsx (hunk 6 above)
Comment (Nokel81) [reply src/renderer/bootstrap.tsx]: Nope. Didn't remember that I did this.

**Review by Nachasic — DISMISSED** [review]

**Review by Nachasic — APPROVED** [review]
